agent/engine/execcmdagent/manager_linux.go

```diff
@@ -0,0 +1,209 @@
+// +build linux
+
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+package execcmdagent
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/cihub/seelog"
+	"github.com/docker/docker/api/types"
+
+	apicontainer "github.com/aws/amazon-ecs-agent/agent/api/container"
+	apitask "github.com/aws/amazon-ecs-agent/agent/api/task"
+	"github.com/aws/amazon-ecs-agent/agent/dockerclient"
+	"github.com/aws/amazon-ecs-agent/agent/dockerclient/dockerapi"
+	"github.com/aws/amazon-ecs-agent/agent/utils/retry"
+)
+
+// RestartIfStopped restarts the ExecCommandAgent in the container passed as parameter, only for ExecCommandAgent-enabled tasks.
+// The status of the ExecCommandAgent in the container is retrieved using a docker exec inspect call, using the dockerExecID
+// stored in container.GetExecCommandAgentMetadata().DockerExecID.
+//
+// If container.GetExecCommandAgentMetadata() is nil, or the ExecCommandAgent is still running, no action is taken.
+// To actually restart the ExecCommandAgent, this function invokes this instance's Start method.
+func (m *manager) RestartIfStopped(ctx context.Context, client dockerapi.DockerClient, task *apitask.Task, container *apicontainer.Container, containerId string) (bool, error) {
+	if !task.IsExecCommandAgentEnabled() {
+		return false, nil
+	}
+	execMD := container.GetExecCommandAgentMetadata()
+	if execMD == nil { // This means m.Start has never been invoked
+		return false, nil
+	}
+	resp, err := client.InspectContainerExec(ctx, execMD.DockerExecID, dockerclient.ContainerExecInspectTimeout)
+	if err != nil {
+		if _, ok := err.(*dockerapi.DockerTimeoutError); ok {
+			return false, err
+		}
+		// TODO: [ecs-exec] consider retry/adding an extra docker exec inspect API call to get more info
+		return false, err
+	}
+	if resp.Running { // Exec cmd agent is running, nothing to do
+		return false, nil
+	}
+	// Restart if not running
+	//TODO: [ecs-exec] retry only for certain exit codes?
+	seelog.Warnf("Task engine [%s]: ExecCommandAgent Process stopped (exitCode:=%d) for container %s, restarting...", task.Arn, resp.ExitCode, containerId)
```

Comment:
is this the exit code of ExecCommandAgent process or of the container? Is there documentation or design around different exit codes from ExecCommandAgent?

Reply:
If there isn't probably better to remove the TODO, as it implies that ExecCommandAgent has logic around it's exit codes

Reply:
I prefer to leave the TODO here, as there are discussions in progress about the exit codes. Before this feature branch is merged, all the `[ecs-exec]` will be addressed, either by removing them or by implementing additional stuff.

Reply:
Oh, and the exit code is the exit code of ExecCommandAgent